dartagnan/src/main/java/com/dat3m/dartagnan/parsers/cat/VisitorBase.java

```diff
@@ -0,0 +1,347 @@
+package com.dat3m.dartagnan.parsers.cat;
+
+import com.dat3m.dartagnan.exception.ParsingException;
+import com.dat3m.dartagnan.parsers.CatBaseVisitor;
+import com.dat3m.dartagnan.parsers.CatParser.*;
+import com.dat3m.dartagnan.program.filter.FilterAbstract;
+import com.dat3m.dartagnan.program.filter.FilterBasic;
+import com.dat3m.dartagnan.program.filter.FilterIntersection;
+import com.dat3m.dartagnan.program.filter.FilterMinus;
+import com.dat3m.dartagnan.program.filter.FilterUnion;
+import com.dat3m.dartagnan.wmm.Definition;
+import com.dat3m.dartagnan.wmm.Relation;
+import com.dat3m.dartagnan.wmm.Wmm;
+import com.dat3m.dartagnan.wmm.axiom.Axiom;
+import com.dat3m.dartagnan.wmm.definition.*;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.dat3m.dartagnan.program.event.Tag.VISIBLE;
+import static com.dat3m.dartagnan.wmm.relation.RelationNameRepository.ID;
+import static com.google.common.base.Verify.verify;
+import static java.util.Objects.requireNonNullElseGet;
+
+class VisitorBase extends CatBaseVisitor<Object> {
+
+    private final Wmm wmm;
+    private final Map<String, Object> namespace = new HashMap<>();
+    private Relation current;
+    private String alias;
+    private final Map<String, Relation> termMap = new HashMap<>();
+
+    VisitorBase() {
+        this.wmm = new Wmm();
+    }
+
+    @Override
+    public Object visitMcm(McmContext ctx) {
+        super.visitMcm(ctx);
+        return wmm;
+    }
+
+    @Override
+    public Void visitAxiomDefinition(AxiomDefinitionContext ctx) {
+        try {
+            Relation r = relation(ctx.e);
+            Constructor<?> constructor = ctx.cls.getConstructor(Relation.class, boolean.class, boolean.class);
+            Axiom axiom = (Axiom) constructor.newInstance(r, ctx.negate != null, ctx.flag != null);
+            if (ctx.NAME() != null) {
+                axiom.setName(ctx.NAME().toString());
+            }
+            wmm.addAxiom(axiom);
+        } catch (NoSuchMethodException | InstantiationException | IllegalAccessException |
+                 InvocationTargetException e) {
+            throw new ParsingException(ctx.getText());
+        }
+        return null;
+    }
+
+    @Override
+    public Void visitLetDefinition(LetDefinitionContext ctx) {
+        String n = ctx.n.getText();
+        alias = ctx.n.getText();
+        Object o = ctx.e.accept(this);
+        namespace.put(n, o);
```

Comment:
I would name the local variables properly. I assume "n" stands for "name" but what does "o" stand for? "Object"?
That is too non-descriptive. You can call it "definedPred" or simply "pred" (I prefer the longer version) to at least indicate that we define a predicate (unary or binary) here.

Also, you can put `alias = n` instead of reusing the context (whose `getText` method has a non-descriptive name).